### ISSUE TITLE
extensible audits

### ISSUE THREAD

**Issue by marcacohen** [GoogleChrome/lighthouse#523]
Had a convo w/ Paul Lewis over breakfast and it dawned on me that instead of trying to compile and maintain the canonical set of audits for everyone ever, it might be more effective to define an extensible interface for adding audits to Lighthouse.

Example: I'm interested in a family of apps that do aggressive pre-fetching of content so that you can view pages you've never visited. I'd like to create an audit that fetches the main landing page, disables network connectivity, iterates all featured links, and validates that each one returns some non-trivial content and a 200 OK. 

The interface for these extensions should allow me to provide a specification file (e.g. lighthouse-ext.json) that encodes several things, like:

- the name of my audit
- short description of what it does
- some way of defining the data it needs from the lighthouse core to perform its audit (e.g. parsed dom tree, interval based screenshots, raw perf data, interactive access to the running page)
- whether it can run in parallel with other extension audits

The run time interface would involve the lighthouse core calling each extension audit (parallelizing wherever possible) and passing the requested data (typically the parsed dom tree). The extension would run its audit(s) and return a set of scores along with a list of html or json formatted findings (for easy presentation), which the core would integrate into the extensions section of the more comprehensive report.

The extensions could be implemented as node modules, distributed via npm, with certain required additional components (e.g. the extension descriptor file). By cataloging these extensions, the lighthouse landing page can provide a search function for finding and invoking user selected extensions on a per audit basis.

The best reason of all to adopt this approach, IMHO, is that it will likely encourage more people to participate and hopefully build up a whole ecosystem around Lighthouse.

**Comment by paullewis** [comment]
I'm very supportive of this. We do support custom configs at the moment, but there's no feature (yet) for adding custom gatherers and audits, which would both be needed for this to work. We'd also need to define & document the APIs and so forth that the custom pieces can use, none of which are blockers of course.

Short version: I'm +1'ing this idea, I'm just not certain what the timeline looks like.

**Comment by blaine** [comment]
+1 I would *love* to see this, and agree with all of the motivations. For what it's worth, I looked at doing this after the Amsterdam PWA conference, and (for obvious reasons) didn't dive right in. I've been on vacation / traveling for a few weeks, and I'm just getting back online now – super happy to see this in here.

If I can be helpful towards getting this going, I'm happy to write some code.

For what it's worth, the approach that I've imagined so far would look something like a repository /  project per test suite (either alongside an app, or standalone), with npm modules included for publicly available tests, and the ability to have a local set of tests, too.

Setup would be something like:
```
$ lighthouse create myproject [path_to_project]
$ cd path_to_project
$ npm install lighthouse-pwa-tests
$ npm install me/my-own-common-package
$ npm run lighthouse [$url]

Where $url might be configurable, but just like the existing workflow, should be specifiable via the command line. Lighthouse would look at the current path, maybe for a lighthouse.conf file, and start including relevant modules and running either all available audits, or just those specified in the config.

My ultimate goal for this is a low-cost, fine-grained, automated, and reproducible set of performance, accessibility, and feature tests to support both our performance work and our PWA work. I've heard rumours that there's a working Chrome headless browser that would make this the basis for an amazing test suite for us at Condé Nast, and I'm sure many others. If we can help kick-start this and support writing some audits, all the better!

**Comment by paulirish** [comment]
Hey blaine. Yeah sounds good. 

I think the setup you have sketched out here is mostly what we've been thinking too.

I don't think we're ready to boot our built-in audits to separate modules, but we can definitely start exploring how to register audits + gatherers that are pulled in from NPM. 

If you're interested, I think it'd be great to dig in here. :)

really valuable to explore this some more. Maybe you could start poking around the code and flesh out a proposal for integration? (Or a strawman implementation).

We could start with either 

1. A proposal for the API. Could include the overall user workflow and an idea of how we implement it on our side. And what the user configuration looks like. Could post it here in this issue.
2. A strawman implementation of the same.

Either works for me, though I'm sure we'll all have some discussion on it. When you're ready we can have you join our regular meeting to talk it out in voice a bit more.

**Comment by brendankenny** [comment]
The good news is that we've been relatively disciplined about moving toward this design(ish) since the beginning, so I believe we're only a few small changes away from it. A path from here to there could be something as simple as:
- the config already allows specifying gatherers and audits by name. Expand this to something like the node [`require` algorithm](https://nodejs.org/api/modules.html#modules_all_together): if it's a built-in gatherer or audit, it can be referred to by name, if not, use the file found by `require.resolve`. That would require primary changes [here](https://github.com/GoogleChrome/lighthouse/blob/0269917f968214a8ac58b26246bf0927998af133/lighthouse-core/config/index.js#L177) and [here](https://github.com/GoogleChrome/lighthouse/blob/0269917f968214a8ac58b26246bf0927998af133/lighthouse-core/gather/gather-runner.js#L237), and then smaller changes around things we have to do anyways, like making sure audit whitelisting makes sense with a custom config, figuring out how to clearly handle audits that require artifacts not found in imported artifacts, etc.

That *should* be all that's needed to allow extensions added by users from an arbitrary location (npm modules, manual checkout somewhere, etc) and have all the other flexibility mentioned above without overcomplicating the default case or leaving us with a bunch of app architecture we don't need.

On the documentation side:
- any requirements of an audit (like required artifacts) should already by covered by the audit meta data and handled by lighthouse. Any aspects missed by a user (like an audit that requires a gatherer that wasn't added) should complain loudly in a user fixable way.
- We need to document the [gatherer](https://github.com/GoogleChrome/lighthouse/blob/0269917f968214a8ac58b26246bf0927998af133/lighthouse-core/gather/gatherers/gatherer.js) and [audit](https://github.com/GoogleChrome/lighthouse/blob/0269917f968214a8ac58b26246bf0927998af133/lighthouse-core/audits/audit.js) base classes to make it clear what each method and property does when/if overridden. Good news is we already have many clear examples of each since core gatherers/audits are written exactly as an extension gatherer/audit would be written :)
- We need to document the gatherer scheduling process, as this is still a very manual process. Easy to schedule one of your js-dependent gatherers while another gatherer has disabled javascript right now.

**Comment by hsablonniere** [comment]
> it might be more effective to define an extensible interface for adding audits to Lighthouse.

Yeah, definitely!

> and hopefully build up a whole ecosystem around Lighthouse.

Looking at this tool in the details (source code, issues and PRs) for a few days, I think this is really promising. Lighthouse's baseline talks about PWA and yes we need tools to push the PWA concept and make it known and easy to access. But as a newcomer, looking a the gatherer/audit/aggregator architecture, I really see a broader performance swiss army knife.

> custom gatherers and audits

The title of this issue should also mention gatherers.

> I've heard rumours that there's a working Chrome headless browser that would make this the basis for an amazing test suite for us at Condé Nast, and I'm sure many others.

So true. It's a bit cumbersome to have the UI sitting on my laptop and I assume CI tools would love it.

> We need to document the gatherer scheduling process, as this is still a very manual process. Easy to schedule one of your js-dependent gatherers while another gatherer has disabled javascript right now.

Again, as a new comer it was hard to find out how the tool works internally and which part serves what purpose. I know such documentation was not really needed yet but if we provide a way to add custom gatherer and audits this will be important. I may help on this matter if you're interested.

### About the proposed solutions to achieve this

I think tools like eslint and their plugin systems could be a good model? I'm not sure of how it works in the details but when I use eslint, I can easily NPM-install a lodash plugin and configure additional rules. It works for the CLI tool and with the module directly.

**Comment by paulirish** [comment]
fixed by #593

basic docs added in 8d696afa45a06bba2a678c2ad0ecc07be021c816
